File: Lab_01/Java/src/lab_01/Polynomial.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package lab_01;

/**
 *
 * @author Nyefan
 */
public class Polynomial {
  
  private static int MAX_DEGREE = 20;
  private double[] coefficients = new double[MAX_DEGREE]; 
  private int length;
  private String printable = "";
  private double sum = 0;
  
  public Polynomial(double[] coefficients) throws IllegalArgumentException {
    if (coefficients.length > MAX_DEGREE) {
      throw new IllegalArgumentException("A Polynomial cannot be initialized with a degree greater than 20;");
    }
    this.length = coefficients.length;
    System.arraycopy(coefficients, 0, this.coefficients, 0, this.length);
  }
  
  public Polynomial add(Polynomial p2, boolean toPrint) {
    double [] tempCoeff = this.coefficients;
    for(int i = 0; i < Math.max(this.length, p2.length); i++) {
      tempCoeff[i] += p2.coefficients[i];
    }
    
    Polynomial returnable = new Polynomial(tempCoeff);
    
    if (toPrint) {
      System.out.print("\nThe sum of the polynomials,\n ");
      print(); System.out.print(" and "); p2.print();
      System.out.print(" is ");
      returnable.print();
    }
    return this;
  }
  
  public Polynomial differentiate(boolean toPrint) {
    double[] coeff = new double[this.length-1];
    for(int i = 0; i < this.length-1; i++) {
      coeff[i] = (i+1)*coefficients[i+1];
    }
    
    Polynomial returnable = new Polynomial(coeff);
    
    if (toPrint) {
      System.out.print("\nThe derivative of the polynomial,\n ");
      print();
      System.out.print(" is ");
      returnable.print();
    }
    
    return returnable;
  }
  
  public double evaluate(double x, boolean toPrint) {
    if (sum == 0) {
      for(int i = 0; i < this.length; i++) {
        sum += coefficients[i]*Math.pow(x,i);
      }
    }
    if (toPrint) {
      System.out.print("\nThe polynomial,\n ");
      print();
      System.out.println(" evaluated at " + x + "\n is " + formatDouble(sum) + ".");
    }
    return sum;
  }
  
  public void print() {
    if (!printable.equals("")) {
      System.out.println(printable);
    } else {
      parseCoefficients();
      System.out.println(printable);
    }
  }
  
  private double formatDouble(double x) {
    return ((double)( Math.round(x*10000)))/10000;
  }
  
  private void parseCoefficients() {
    printable = "";
    printable += formatDouble(coefficients[0]);
    for(int i = 1; i < length-1; i++) {
      if(this.coefficients[i] != 0) {
        printable += " + " + formatDouble(this.coefficients[i]) + "x^" + i;
      }
    }
  }
}

File: Lab_01/C++/Polynomial.cpp
#include "Polynomial.h"

/** 
 * 
 */
Polynomial::Polynomial() : length(MAX_DEGREE), printable(""), PRINT_PRECISION(5), sum(0) {
    std::fill(this->coefficients, this->coefficients + MAX_DEGREE, 0);
}

/** 
 * 
 */
Polynomial::Polynomial(double coefficients[], int order = MAX_DEGREE-1) : length(order+1), printable(""), PRINT_PRECISION(5), sum(0) {
    std::fill(this->coefficients, this->coefficients + MAX_DEGREE, 0);
    for (int i = 0; i < length; i++) {
        this->coefficients[i] = coefficients[i];
    }
}

/** 
 * 
 */
Polynomial Polynomial::add(Polynomial p2, const bool toPrint = 0) {
    double tempCoeff[MAX_DEGREE] = {0};
    
    for(int i = 0; i < std::max(this->length, p2.length); i++) {
        tempCoeff[i] += p2.coefficients[i] + coefficients[i];
    }

    Polynomial returnable(tempCoeff, std::max(this->length, p2.length)-1);

    if (toPrint) {
        std::cout << "\nThe sum of the polynomials,\n ";
        print(); std::cout << " and "; p2.print();
        std::cout << " is "; returnable.print();
    }

    return returnable;
}

/** 
 * 
 */
Polynomial Polynomial::differentiate(const bool toPrint = 0) {
    double tempCoeff[MAX_DEGREE] = {0};

    for(int i = 0; i < this->length - 1; i++) {
        tempCoeff[i] = (i+1)*this->coefficients[i+1];
    }

    Polynomial returnable(tempCoeff, this->length-1);

    if (toPrint) {
        std::cout << "\nThe derivative of the polynomial,\n ";
        print(); std::cout << " is ";
        returnable.print();
    }

    return returnable;
}

/** 
 * 
 */
double Polynomial::evaluate(const double x, const bool toPrint = 0) {
    if (sum == 0) {
        for (int i = 0; i < this->length; i++) {
            sum += this->coefficients[i]*std::pow(x,i);
        }
    }

    if (toPrint) {
        std::cout << "\nThe polynomial,\n ";
        print();
        std::cout << std::setprecision(PRINT_PRECISION) << " evaluated at " << x << "\n is " << sum << ".\n";
    }

    return sum;
}

/** 
 * 
 */
void Polynomial::print() {

    int counter = 1;
    
    if (this->coefficients[0] != 0) {
        std::cout << std::setprecision(PRINT_PRECISION) << this ->coefficients[0] << " + ";
    } 
    while (this->coefficients[counter] == 0) {
        ++counter;
    } std::cout << std::setprecision(PRINT_PRECISION) << this->coefficients[counter] << "x^" << counter;
    ++counter;
    for (; counter < length; counter++) {
        if (this->coefficients[counter] != 0) {
            std::cout << std::setprecision(PRINT_PRECISION) << " + " << this->coefficients[counter] << "x^" << counter;
        }
    } 
    if (counter < length && this->coefficients[counter] != 0) {
        std::cout << std::setprecision(PRINT_PRECISION) << this->coefficients[length] << "x^" << length;
    } std::cout << '\n';
}

